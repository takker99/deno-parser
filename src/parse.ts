import { skip } from "./skip.ts";
import { eof } from "./eof.ts";
import type { Expected } from "./expected.ts";
import type { SourceLocation } from "./SourceLocation.ts";

/**
 * The parsing action.
 * This should only be called directly when writing custom parsers.
 *
 * > [!NOTE]
 * > Make sure to use {@linkcode merge} when combining multiple
 * > {@linkcode ActionResult}s or else you will lose important parsing information.
 *
 * @param context A parsing context
 * @returns An {@linkcode ActionResult}
 */
export type Parser<A, Input extends ArrayLike<unknown>> = (
  input: Input,
  location?: SourceLocation,
  options?: ParserOptions,
) => ParseResult<A>;

// deno-lint-ignore no-empty-interface
export interface ParserOptions {}

/** Extracts the parsed type from a {@linkcode Parser}. */
export type ParserResult<
  Input extends ArrayLike<unknown>,
  P extends Parser<unknown, Input>,
> = P extends Parser<infer A, Input> ? A
  : never;

/**  Extracts the input type from a {@linkcode Parser}. */
export type ParserInput<P extends Parser<unknown, ArrayLike<unknown>>> =
  P extends Parser<unknown, infer I extends ArrayLike<unknown>> ? I : never;

export type ParseResult<A> = ParseOk<A> | ParseFail;

/**
 * Represents a successful parse result.
 */
export interface ParseOk<A> {
  /**
   * Whether the parse was successful.
   */
  ok: true;

  /** The parsed value */
  value: A;

  /** List of expected values at the location the parse failed */
  expected: Expected[];

  /** The next location where a parse starts to read */
  next: SourceLocation;
}

/**
 * Represents a failed parse result, where it failed, and what types of
 * values were expected at the point of failure.
 */
export interface ParseFail {
  /**
   * Whether the parse was successful.
   */
  ok: false;

  /** List of expected values at the location the parse failed */
  expected: Expected[];
}

export const isOk = <A>(
  parseResult: ParseResult<A>,
): parseResult is ParseOk<A> => parseResult.ok;

export const isFail = <A>(
  parseResult: ParseResult<A>,
): parseResult is ParseFail => !parseResult.ok;

/**
 * Returns a parse result with either the value or error information.
 *
 * Parses the entire `input` string, returning a {@linkcode ParserResult} with
 * the parse value if successful, otherwise a failure value indicating where the
 * error is and what values we were looking for at the time of failure.
 *
 * > [!NOTE]
 * > `parse` assumes you are parsing the entire input and will fail unless you do so.
 *
 * @example
 * ```ts
 * import { text, parse } from "@takker/parser";
 * import { assertEquals } from "@std/assert";
 *
 * const a = text("a");
 * const result1 = parse(a, "a");
 * if (result1.ok) {
 *   assertEquals(result1.value, "a");
 * } else {
 *   const { expected } = result1;
 *   console.error(`${expected.length} parse error(s) occurred:\n${
 *     expected.map(({ expected, location: { line, column } }) =>
 *       `at line ${line} column ${column}: expected ${[...expected].join(", ")}`
 *     )
 *   }`);
 * }
 * ```
 */
export const parse = <A, Input extends ArrayLike<unknown>>(
  parser: Parser<A, Input>,
  input: Input,
): ParseResult<A> => skip(parser, eof)(input);

/**
 * Returns the parsed result or throws an error.
 *
 * Return the result from successfully parsing the `input` string.
 *
 * If the parse fails, throws an {@linkcode Error} with a message describing the line/column
 * and what values were expected. This method is provided for convenience in case
 * you are not interested in handling the failure case.
 *
 * > [!NOTE]
 * > If you plan on handling failures, use {@linkcode parse} directly to get full
 * > information about went wrong so you can present the error message better for our
 * > application.
 *
 * ```ts
 * import { text, tryParse } from "@takker/parser";
 * import { assertEquals } from "@std/assert";
 *
 * const a = text("a");
 * assertEquals(tryParse(a, "a"), "a");
 * ```
 */
export const tryParse = <A, Input extends ArrayLike<unknown>>(
  parser: Parser<A, Input>,
  input: Input,
): A => {
  const result = parse(parser, input);
  if (isOk(result)) return result.value;
  const { expected } = result;
  const message = `${expected.length} parse error(s) occurred:\n${
    expected.map(({ expected, location: { line, column } }) =>
      `at line ${line} column ${column}: expected ${[...expected].join(", ")}`
    )
  }`;
  throw new Error(message);
};
