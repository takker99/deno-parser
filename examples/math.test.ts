import { assertEquals, assertThrows } from "@std/assert";
import {
  type MathExpr,
  MathOperator1,
  MathOperator2,
  SimpleMath,
} from "./math.ts";
import { assertSnapshot } from "@std/testing/snapshot";
import { parse, tryParse } from "../src/parse.ts";

Deno.test("2 part expression", async (t) => {
  const text = `2 * 3 + 4`;
  const expr = tryParse(SimpleMath, text);
  await assertSnapshot(t, parse(SimpleMath, text));
  assertEquals(expr.calculate(), 10);
  assertEquals(`${expr}`, `((2 * 3) + 4)`);
});

Deno.test("2 part expression with parentheses", async (t) => {
  const text = `2 * (3 + 4)`;
  const expr = tryParse(SimpleMath, text);
  await assertSnapshot(t, parse(SimpleMath, text));
  assertEquals(expr.calculate(), 14);
  assertEquals(`${expr}`, `(2 * (3 + 4))`);
});

Deno.test("unary and binary minus", async (t) => {
  const text = `-1 - -1`;
  const expr = tryParse(SimpleMath, text);
  await assertSnapshot(t, parse(SimpleMath, text));
  assertEquals(expr.calculate(), 0);
  assertEquals(`${expr}`, `((- 1) - (- 1))`);
});

Deno.test("large expression", async (t) => {
  const text = `-2 + 3 * 4 - 5 / 7 ** 6`;
  const expr = tryParse(SimpleMath, text);
  await assertSnapshot(t, parse(SimpleMath, text));
  assertEquals(expr.calculate(), 9.999957500701239);
});

Deno.test("MathOperator1", () => {
  const expr: MathExpr = {
    calculate() {
      return 0;
    },
  };
  assertThrows(() => new MathOperator1("bad", expr).calculate());
});

Deno.test("MathOperator2", () => {
  const expr: MathExpr = {
    calculate() {
      return 0;
    },
  };
  assertThrows(() => new MathOperator2("bad", expr, expr).calculate());
});
