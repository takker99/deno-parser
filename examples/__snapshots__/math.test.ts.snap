export const snapshot = {};

snapshot[`2 part expression 1`] = `
{
  type: "ParseOK",
  value: MathOperator2 {
    left: MathOperator2 {
      left: MathNumber {
        value: 2,
      },
      operator: "*",
      right: MathNumber {
        value: 3,
      },
    },
    operator: "+",
    right: MathNumber {
      value: 4,
    },
  },
}
`;

snapshot[`2 part expression with parentheses 1`] = `
{
  type: "ParseOK",
  value: MathOperator2 {
    left: MathNumber {
      value: 2,
    },
    operator: "*",
    right: MathOperator2 {
      left: MathNumber {
        value: 3,
      },
      operator: "+",
      right: MathNumber {
        value: 4,
      },
    },
  },
}
`;

snapshot[`unary and binary minus 1`] = `
{
  type: "ParseOK",
  value: MathOperator2 {
    left: MathOperator1 {
      expression: MathNumber {
        value: 1,
      },
      operator: "-",
    },
    operator: "-",
    right: MathOperator1 {
      expression: MathNumber {
        value: 1,
      },
      operator: "-",
    },
  },
}
`;

snapshot[`large expression 1`] = `
{
  type: "ParseOK",
  value: MathOperator2 {
    left: MathOperator2 {
      left: MathOperator1 {
        expression: MathNumber {
          value: 2,
        },
        operator: "-",
      },
      operator: "+",
      right: MathOperator2 {
        left: MathNumber {
          value: 3,
        },
        operator: "*",
        right: MathNumber {
          value: 4,
        },
      },
    },
    operator: "-",
    right: MathOperator2 {
      left: MathNumber {
        value: 5,
      },
      operator: "/",
      right: MathOperator2 {
        left: MathNumber {
          value: 7,
        },
        operator: "**",
        right: MathNumber {
          value: 6,
        },
      },
    },
  },
}
`;
